packages/dom-adapters/src/utils/helpers.ts

```diff
@@ -0,0 +1,325 @@
+      return [child, totalLength];
+    }
+
+    totalLength += getLength(child);
+  }
+
+
+  /**
+   * This is unreachable code in normal operation, but we need it to have consistent types
+   */
+  /* Stryker disable next-line StringLiteral */
+  /* istanbul ignore next */
+  throw new Error(`Child is not found by ${index} index`);
+}
+
+/**
+ * Returns DOM Range by char range (relative to input)
+ *
+ * @param input - input element char indexes are related to
+ * @param start - start char index
+ * @param end - end char index
+ */
+export function getRange(input: HTMLElement, start: number, end: number): Range {
+  const length = getLength(input);
+
+  if (start < 0 || start > length || end < 0 || end > length) {
+    throw new Error('InlineToolAdapter: range is out of bounds');
+  }
+
+  let startContainer: Node = input;
+  let startOffset = start;
+  let endContainer: Node = input;
+  let endOffset = end;
+
+  /**
+   * Find start node and offset for the range
+   */
+  while (!(startContainer instanceof Text)) {
+    const [child, offset] = findChildByIndex(startContainer, startOffset);
+
+    startContainer = child;
+    startOffset = startOffset - offset;
+  }
+
+  /**
+   * Find end node and offset for the range
+   */
+  while (!(endContainer instanceof Text)) {
+    const [child, offset] = findChildByIndex(endContainer, endOffset);
+
+    endContainer = child;
+    endOffset = endOffset - offset;
+  }
+
+  const range = new Range();
+
+  /**
+   * If startOffset equals to the length of the startContainer, we need to set start after the startContainer
+   *
+   * However, we also need to consider parent nodes siblings
+   */
+  if (startOffset === getLength(startContainer)) {
+    let nextSibling = startContainer.nextSibling;
+    let parent = startContainer.parentNode!;
+
+    while (nextSibling === null && parent !== input) {
+      nextSibling = parent.nextSibling;
+      parent = parent.parentNode!;
+    }
+
+    if (nextSibling !== null) {
+      range.setStartBefore(nextSibling);
+    } else {
+      range.setStartAfter(startContainer);
+    }
+  } else {
+    range.setStart(startContainer, startOffset);
+  }
+
+  /**
+   * If endOffset equals to 0, we need to set end before the endContainer
+   */
+  if (endOffset === 0) {
+    range.setEndBefore(endContainer);
+  /**
+   * If endOffset equals to the length of the endContainer, we need to set end after the endContainer
+   *
+   * We need to consider parent nodes siblings as well
+   */
+  } else if (endOffset === getLength(endContainer)) {
+    let nextSibling = endContainer.nextSibling;
+    let parent = endContainer.parentNode!;
+
+    while (nextSibling === null && parent !== input) {
+      nextSibling = parent.nextSibling;
+      parent = parent.parentNode!;
+    }
+
+    if (nextSibling !== null) {
+      range.setEndBefore(nextSibling);
+    } else {
+      range.setEndAfter(endContainer);
+    }
+  } else {
+    range.setEnd(endContainer, endOffset);
```

Comment:
seems like this code could me moved to some method or util since it repeated above